// lightning-flow-scanner-core/out/main/rules/ProcessBuilder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProcessBuilder = void 0;
const core = __importStar(require("../internals/internals"));
const RuleCommon_1 = require("../models/RuleCommon");
class ProcessBuilder extends RuleCommon_1.RuleCommon {
    constructor() {
        super({
            name: 'ProcessBuilder',
            label: 'No Process Builder',
            description: "Salesforce is transitioning away from Workflow Rules and Process Builder in favor of Flow. Ensure you're prepared for this transition by migrating your organization's automation to Flow. Refer to official documentation for more information on the transition process and tools available.",
            supportedTypes: core.FlowType.processBuilder,
            docRefs: [{ 'label': 'Process Builder Retirement', 'path': 'https://help.salesforce.com/s/articleView?id=000389396&type=1' }],
            isConfigurable: true,
            autoFixable: false
        });
    }
    execute(flow, options) {
        return new core.RuleResult(this, [new core.ResultDetails(new core.FlowAttribute('Workflow', "processType", '== Workflow'))]);
    }
}
exports.ProcessBuilder = ProcessBuilder;
