// lightning-flow-scanner-core/out/main/store/DefaultRuleStore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultRuleStore = void 0;
const APIVersion_1 = require("../rules/APIVersion");
const AutoLayout_1 = require("../rules/AutoLayout");
const CopyAPIName_1 = require("../rules/CopyAPIName");
const DMLStatementInLoop_1 = require("../rules/DMLStatementInLoop");
const DuplicateDMLOperation_1 = require("../rules/DuplicateDMLOperation");
const FlowDescription_1 = require("../rules/FlowDescription");
const FlowName_1 = require("../rules/FlowName");
const HardcodedId_1 = require("../rules/HardcodedId");
const InactiveFlow_1 = require("../rules/InactiveFlow");
const MissingFaultPath_1 = require("../rules/MissingFaultPath");
const MissingNullHandler_1 = require("../rules/MissingNullHandler");
const ProcessBuilder_1 = require("../rules/ProcessBuilder");
const SOQLQueryInLoop_1 = require("../rules/SOQLQueryInLoop");
const UnconnectedElement_1 = require("../rules/UnconnectedElement");
const UnusedVariable_1 = require("../rules/UnusedVariable");
exports.DefaultRuleStore = {
    APIVersion: APIVersion_1.APIVersion,
    AutoLayout: AutoLayout_1.AutoLayout,
    CopyAPIName: CopyAPIName_1.CopyAPIName,
    DMLStatementInLoop: DMLStatementInLoop_1.DMLStatementInLoop,
    DuplicateDMLOperation: DuplicateDMLOperation_1.DuplicateDMLOperation,
    FlowDescription: FlowDescription_1.FlowDescription,
    FlowName: FlowName_1.FlowName,
    HardcodedId: HardcodedId_1.HardcodedId,
    MissingFaultPath: MissingFaultPath_1.MissingFaultPath,
    MissingNullHandler: MissingNullHandler_1.MissingNullHandler,
    ProcessBuilder: ProcessBuilder_1.ProcessBuilder,
    SOQLQueryInLoop: SOQLQueryInLoop_1.SOQLQueryInLoop,
    UnconnectedElement: UnconnectedElement_1.UnconnectedElement,
    UnusedVariable: UnusedVariable_1.UnusedVariable,
    InactiveFlow: InactiveFlow_1.InactiveFlow,
};
